src/styles/MuiThemeProvider.js

```diff
@@ -1,20 +1,25 @@
 import {Component, PropTypes} from 'react';
 import getMuiTheme from './getMuiTheme';
+import getUniqueIdGenerator from '../utils/getUniqueIdGenerator';
+
 
 class MuiThemeProvider extends Component {
 
   static propTypes = {
     children: PropTypes.element,
     muiTheme: PropTypes.object.isRequired,
+    uniqueIdGen: PropTypes.func,
   };
 
   static childContextTypes = {
     muiTheme: PropTypes.object.isRequired,
+    uniqueIdGen: PropTypes.func.isRequired,
   };
 
   getChildContext() {
     return {
       muiTheme: this.props.muiTheme || getMuiTheme(),
+      uniqueIdGen: this.props.uniqueIdGen || getUniqueIdGenerator(),
```

Comment:
@oliviertassinari This method attaches a new generator(call `getUniqueIdGenerator`) to each `<MuiThemeProvider>`, so children of `<MuiThemeProvider>` will use the generator in its context to create uniqueId, instead of a shared global  generator.